### PR TITLE
Add style document

### PR DESCRIPTION
**Pull request by chris-taylor-ntt**
Adds a style guide for `launch-cli`. This is primarily driven by the items I logged in AzDO after we merged the latest set of commands and sets a baseline for contributions to the repository going forward.

### CODE REVIEW

**Comment by chris-taylor-ntt** [comment]
I suggest reviewers click on the kebab next to the raw Markdown and use "View File" to review so that you get the formatted view.

File: STYLE.md [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+Prefer simple types, but when complex types are required, you are encouraged to define them as a standalone type at the top of the module and then use them throughout in function and variable annotations.
+
+```py
+# Readability suffers as the number of arguments and their complexity grows:
+def handle_bar(bar: dict[str, set[str]]) -> dict[str, set[str]]:
+    output: dict[str, set[str]] = {}
+    ...
+
+# Do this instead:
+MyComplexType = dict[str, set[str]]
```
Comment (aaron-launch) [inline STYLE.md:48]: Definitions, enums and constants should be defined in a separate file and imported in.

File: STYLE.md [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+
+### Outputs
+
+Never `print()`, use `click.echo()` and `click.secho()` from a top-level command if you need to write text directly to the terminal.
+
+When using `click.secho()`, the following styles are recommended:
+
+```py
+click.secho("Denotes a failure, text will appear red on most terminals", fg="red")
+click.secho("Denotes a warning, yellow on most terminals", fg="yellow")
```
Comment (aaron-launch) [inline STYLE.md:70]: Color is not available in logs, does this prefix this with a [ERROR] or [WARN] ?

File: STYLE.md [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+
+When mocking objects, prefer the `pytest`-style mocks over `unittest`-style decorated or context manager mocks for three reasons:
+
+1. Decorators are zero-reuse, they must be copy-pasted around
+2. Large numbers of decorators create visual clutter that makes code more difficult to read
+3. `pytest` eschews context manager mocks because it provides cleanup guarantees (and reduces indent depth as a bonus!)
+
+```py
+# Avoid:
+@patch(“thing.to.mock”)
```
Comment (aaron-launch) [inline STYLE.md:263]: Cloud providers, like AWS should not be patched through `unittest`. [Moto](https://github.com/nexient-llc/launch-cli.git) should be used for AWS. If there are purpose built mocking for libraries we consume, we should prioritize using those over `unittest patch`.

File: STYLE.md [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+    mocker.patch("thing.specific.to.this.test", side_effect=ValueError)
+    ...
+```
+
+If you need to reuse a series of mocks with a consistent set of side effects, create a fixture and utilize that fixture in your test code. If your side effect requirements are unique to the test function, the side effect statements belong inside the function as it's part of the test code.
+
+When writing tests, reserve the use of decorators for test control flow, like `parametrize`, `skip`, `skipif`, `xfail`, etc.
+
+```py
+@pytest.mark.parametrize("input_a", [1,2,3])
```
Comment (aaron-launch) [inline STYLE.md:293]: In my experience, it has been easier to manage test data by aggregating it in a single folder/file and reading the data in through a `fixture` to be used by the tests.

File: STYLE.md (hunk 1 above)
Comment (chris11-taylor-nttd) [reply STYLE.md:48]: I'm largely on board with this, especially for items that have a broad scope. If it's an enum or constant that is intended to be shared across module boundaries, I'm all for it. If it's purely internal to the module I think there's a balance to be struck between including it with the logic versus splitting it off into a separate file.

Do you already have some places in mind where you'd like to perform this split? Is there a good name that captures the union of definitions, enums, and constants that we can codify into this document for the future?

What I want to try to avoid is an expansive file tree populated by tiny files, this can lead to startup performance issues. Python's not terribly quick to initialize in the first place, and extra disk access and calls to importlib don't help especially if you're already paying the bytecode compilation penalty on first run. It's not that noticeable at 10 or even 100 source files but at 1000 you definitely feel it if you don't utilize lazy-loading techniques.

Related: We probably need to spend some time unifying [these constants](https://github.com/nexient-llc/launch-cli/blob/main/src/launch/__init__.py) with the [environment variables](https://github.com/nexient-llc/launch-cli/blob/main/src/launch/env.py) so that we have a consistent interface for overriding default behavior of the software.

File: STYLE.md (hunk 2 above)
Comment (chris11-taylor-nttd) [reply STYLE.md:70]: It doesn't. `secho` is just a "styled echo" so it would be on us to add a tag like [ERROR] or [WARN] by wrapping this in a function that we use throughout our codebase to maintain consistency in our outputs.

Ideally whatever failure we're handling and eventually bubbling up to the user would also be covered by logger statements that provide a timestamp, level, module information, a stack trace, etc. While we might provide a way for end users to adjust logging module verbosity to their liking, for any run through CI/CD we'd want INFO at a minimum and maybe DEBUG in some cases.

I think the next evolution of our logging approach needs to replace the [hardcoded basicConfig](https://github.com/nexient-llc/launch-cli/blob/main/src/launch/cli/entrypoint.py#L39-L43) with something a bit more flexible. We can use fileConfig to pull in an ini-formatted configuration (maybe this lives at ~/.launch-cli/...?) but I've found that JSON or YAML are much more approachable than an ini file, and it's not much more difficult to dump one of those files to a dictionary and feed it to [dictConfig](https://docs.python.org/3/library/logging.config.html#logging.config.dictConfig).

If we do crystallize logging configuration to file, there are some challenges associated with making breaking changes to logging after that point, namely how to reconcile changes to a configuration file with the file that already exists. Do we overwrite the user's settings (and make a backup of the old file), or incrementally update new configs and leave old configs alone? Not necessarily something we need to solve as part of this style guide discussion, just something to think on for the near future.

File: STYLE.md (hunk 3 above)
Comment (chris11-taylor-nttd) [reply STYLE.md:263]: Agreed on avoiding mocking cloud providers directly, it's a frustrating exercise at best. I've never used Moto but it seems like a great place to start. Do you happen to know if there are similar libraries for Azure and GCP?

One other testing library that I've leaned on heavily in the past is [Faker](https://faker.readthedocs.io/en/master/), and while it's not quite like having a fuzzer, I've found having some randomness in test inputs will often expose edge cases and problems that aren't otherwise apparent. This usually results in better code design, more precise test definitions, or both!

At one point Matt and I discussed using [localstack](https://docs.localstack.cloud/user-guide/aws/feature-coverage/) for AWS but I don't remember exactly where we left off on that discussion. I think the main benefit for us would be that we could perform limited e2e-style testing with the tools we're wrapping (TF, TG, maybe others?), but I'm perfectly happy to start with Moto and see where that takes us.

File: STYLE.md (hunk 4 above)
Comment (chris11-taylor-nttd) [reply STYLE.md:293]: There's definitely value in separating our test data from our test code, especially if we intend to use the same test data across multiple test cases.

In my mind, this is a bigger concern with integration and e2e testing than it is with unit tests, since unit tests should be relatively simple and concise by comparison. I find that when unit tests are validating an implementation rather than an outcome (more often than anyone would like to admit) they end up being pretty brittle and often need to change as the implementation changes. 

If we apply this as a broad rule we'll end up reworking not just unit tests, but the fixtures and data files that they rely on as well, amplifying the effort needed for tests. In my experience, if it's difficult to test, people will either just not do the testing or write tests that satisfy coverage but don't perform meaningful validation. Both outcomes are bad.

Do you have a heuristic for when we'd apply this technique? It seems like applying this everywhere leads to file sprawl and an increase in boilerplate that isn't desirable for tests.
